src/bin/pg_autoctl/cli_create_drop_node.c

```diff
@@ -855,3 +864,171 @@ check_nodename(const char *nodename)
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+
+			stop_postgres_and_remove_pgdata_and_config(&mconfig.pathnames,
+													   &mconfig.pgSetup);
+			break;
+		}
+
+		case PG_AUTOCTL_ROLE_KEEPER:
+		{
+			bool missingPgdataIsOk = true;
+			bool pgIsNotRunningIsOk = true;
+			bool monitorDisabledIsOk = true;
+
+			if (!keeper_config_read_file(&config,
+										 missingPgdataIsOk,
+										 pgIsNotRunningIsOk,
+										 monitorDisabledIsOk))
+			{
+				/* errors have already been logged */
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+
+			keeper_cli_destroy_keeper_node(&keeper, &config);
+			break;
+		}
+
+		default:
+		{
+			log_fatal("Unrecognized configuration file \"%s\"",
+					  config.pathnames.config);
+			exit(EXIT_CODE_BAD_CONFIG);
+		}
+	}
+}
+
+
+/*
+ * keeper_cli_destroy_keeper_node destroys a keeper node.
+ */
+void
+keeper_cli_destroy_keeper_node(Keeper *keeper, KeeperConfig *config)
+{
+	/* maybe stop running keeper service first */
+	if (file_exists(config->pathnames.pid))
+	{
+		pid_t pid = 0;
+
+		if (read_pidfile(config->pathnames.pid, &pid))
+		{
+			log_info("An instance of this keeper is running with PID %d, "
+					 "stopping it.", pid);
+
+			if (kill(pid, SIGQUIT) != 0)
+			{
+				log_error("Failed to send SIGQUIT to the keeper's pid %d: %s",
+						  pid, strerror(errno));
+				exit(EXIT_CODE_INTERNAL_ERROR);
+			}
+		}
+	}
+
+	/* only keeper_remove when we still have a state file around */
+	if (file_exists(config->pathnames.state))
+	{
+		bool ignore_monitor_errors = true;
+
+		/* keeper_remove uses log_info() to explain what's happening */
+		if (!keeper_remove(keeper, config, ignore_monitor_errors))
+		{
+			log_fatal("Failed to remove local node from the pg_auto_failover "
+					  "monitor, see above for details");
+
+			exit(EXIT_CODE_BAD_STATE);
+		}
+	}
+	else
+	{
+		log_warn("State file \"%s\" does not exist, skipping keeper remove step",
+				 config->pathnames.state);
+	}
+
+	stop_postgres_and_remove_pgdata_and_config(&(config->pathnames),
+											   &(config->pgSetup));
+}
+
+/*
+ * stop_postgres_and_remove_pgdata_and_config stops PostgreSQL and then removes
+ * PGDATA, and then config and state files.
+ */
+void
+stop_postgres_and_remove_pgdata_and_config(ConfigFilePaths *pathnames,
+										   PostgresSetup *pgSetup)
+{
+	log_info("Stopping PostgreSQL at \"%s\"", pgSetup->pgdata);
+
+	if (!pg_ctl_stop(pgSetup->pg_ctl, pgSetup->pgdata))
+	{
+		log_error("Failed to stop PostgreSQL at \"%s\"", pgSetup->pgdata);
+		exit(EXIT_CODE_PGCTL);
+	}
+
+	/*
+	 * Only try to rm -rf PGDATA if we managed to stop PostgreSQL.
+	 */
+	if (directory_exists(pgSetup->pgdata))
+	{
+		log_info("Removing \"%s\"", pgSetup->pgdata);
+
+		if (!rmtree(pgSetup->pgdata, true))
+		{
+			log_error("Failed to remove directory \"%s\": %s",
+					  pgSetup->pgdata, strerror(errno));
+			exit(EXIT_CODE_INTERNAL_ERROR);
+		}
+	}
+	else
+	{
+		log_info("Skipping removal of \"%s\": directory does not exists",
+				 pgSetup->pgdata);
```

Comment:
Now wondering if that should be a WARN message. It sounds like it should, what do you think?

Reply:
it feels like an unexpected situation, warn might be better